cmd/bee-split/main.go

```diff
@@ -0,0 +1,241 @@
+	host        string // flag variable, http api host
+	port        int    // flag variable, http api port
+	noHttp      bool   // flag variable, skips http api if set
+	ssl         bool   // flag variable, uses https for api if set
+)
+
+// teeStore provides a storage.Putter that can put to multiple underlying storage.Putters
+type teeStore struct {
+	putters []storage.Putter
+}
+
+// newTeeStore creates a new teeStore
+func newTeeStore() *teeStore {
+	return &teeStore{}
+}
+
+// Add adds a storage.Putter
+func (t *teeStore) Add(putter storage.Putter) {
+	t.putters = append(t.putters, putter)
+}
+
+// Put implements storage.Putter
+func (t *teeStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, putter := range t.putters {
+		_, err := putter.Put(ctx, mode, chs...)
+		if err != nil {
+			return []bool{}, err
+		}
+	}
+	return []bool{}, nil
+}
+
+// fsStore provides a storage.Putter that writes chunks directly to the filesystem.
+// Each chunk is a separate file, where the hex address of the chunk is the file name.
+type fsStore struct {
+	path string
+}
+
+// newFsStore creates a new fsStore
+func newFsStore(path string) storage.Putter {
+	return &fsStore{
+		path: path,
+	}
+}
+
+// Put implements storage.Putter
+func (f *fsStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, ch := range chs {
+		chunkPath := filepath.Join(f.path, ch.Address().String())
+		err := ioutil.WriteFile(chunkPath, ch.Data(), 0o666)
+		if err != nil {
+			return []bool{}, err
+		}
+	}
+	return []bool{}, nil
+}
+
+// apiStore provies a storage.Putter that adds chunks to swarm through the HTTP chunk API.
+type apiStore struct {
+	baseUrl string
+}
+
+// newApiStore creates a new apiStor
+func newApiStore(host string, port int, ssl bool) (storage.Putter, error) {
+	scheme := "http"
+	if ssl {
+		scheme += "s"
+	}
+	u := &url.URL{
+		Host:   fmt.Sprintf("%s:%d", host, port),
+		Scheme: scheme,
+		Path:   "bzz-chunk",
+	}
+	return &apiStore{
+		baseUrl: u.String(),
+	}, nil
+}
+
+// Put implements storage.Putter
+func (a *apiStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, ch := range chs {
+		addr := ch.Address().String()
+		buf := bytes.NewReader(ch.Data())
+		url := strings.Join([]string{a.baseUrl, addr}, "/")
+		c := &http.Client{}
+		res, err := c.Post(url, "application/octet-stream", buf)
+		if err != nil {
+			return []bool{}, err
+		}
+		_ = res
+	}
+	return []bool{}, nil
+}
+
+// limitReadCloser wraps the input to the application to limit the input to the given count flag.
+type limitReadCloser struct {
+	io.Reader
+	closeFunc func() error
+}
+
+// newLimitReadCloser creates a new limitReadCloser.
+func newLimitReadCloser(r io.Reader, closeFunc func() error, c int64) io.ReadCloser {
+	return &limitReadCloser{
+		Reader:    io.LimitReader(r, c),
+		closeFunc: closeFunc,
```

Comment:
if you change the struct to the aforementioned suggestion above, you can set `Close` her with the provided `closeFunc`, then you can remove the `Close()` implementation below (if i am not mistaken). just a suggestion

Reply:
I tried that and it didn't work. I think the methods then have to be explicit. Or the object passed in the argument must itself be a readcloser.

Reply:
there you go:
```
diff --git a/cmd/bee-split/main.go b/cmd/bee-split/main.go
index 108be4f..9f52625 100644
--- a/cmd/bee-split/main.go
+++ b/cmd/bee-split/main.go
@@ -125,22 +125,17 @@ func (a *apiStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.C
 // limitReadCloser wraps the input to the application to limit the input to the given count flag.
 type limitReadCloser struct {
        io.Reader
-       closeFunc func() error
+       io.Closer
 }
 
 // newLimitReadCloser creates a new limitReadCloser.
-func newLimitReadCloser(r io.Reader, closeFunc func() error, c int64) io.ReadCloser {
+func newLimitReadCloser(r io.Reader, cl io.Closer, c int64) io.ReadCloser {
        return &limitReadCloser{
-               Reader:    io.LimitReader(r, c),
-               closeFunc: closeFunc,
+               Reader: io.LimitReader(r, c),
+               Closer: cl,
        }
 }
 
-// Close implements io.Closer
-func (l *limitReadCloser) Close() error {
-       return l.closeFunc()
-}
-
 // Split is the underlying procedure for the CLI command
 func Split(cmd *cobra.Command, args []string) (err error) {
        var infile io.ReadCloser
@@ -171,13 +166,13 @@ func Split(cmd *cobra.Command, args []string) (err error) {
                        return err
                }
                defer f.Close()
-               infile = newLimitReadCloser(f, f.Close, inputLength)
+               infile = newLimitReadCloser(f, f, inputLength)
        } else {
                // this simple splitter is too stupid to handle open-ended input, sadly
                if inputLength == 0 {
                        return errors.New("must specify length of input on stdin")
                }
-               infile = newLimitReadCloser(os.Stdin, func() error { return nil }, inputLength)
+               infile = newLimitReadCloser(os.Stdin, newNoopCloser(), inputLength)
        }
 
        // add the fsStore and/or apiStore, depending on flags
@@ -240,3 +235,14 @@ Chunks are saved in individual files, and the file names will be the hex address
                os.Exit(1)
        }
 }
+
+type noopCloser struct {
+}
+
+func newNoopCloser() io.Closer {
+       return noopCloser{}
+}
+
+func (_ noopCloser) Close() error {
+       return nil
+}
```

Reply:
After discussing this in person conclusion is it's ok to skip this.